File: core/src/main/java/y2k/joyreactor/services/requests/AddTagRequest.java
package y2k.joyreactor.services.requests;

import org.jsoup.nodes.Document;
import rx.Observable;
import y2k.joyreactor.common.ObservableUtils;
import y2k.joyreactor.http.HttpClient;

import java.net.URLEncoder;

/**
 * Created by y2k on 19/10/15.
 */
public class AddTagRequest {

    private String tagName;

    public AddTagRequest(String tagName) {
        this.tagName = tagName;
    }

    public Observable<Void> request() {
        return ObservableUtils.create(() -> {
            String tagUrl = "http://joyreactor.cc/tag/" + URLEncoder.encode(tagName);
            Document tagPage = HttpClient.getInstance().getDocument(tagUrl);
            String addTagLink = tagPage.select("a.change_favorite_link").first().absUrl("href");
            HttpClient.getInstance().getText(addTagLink);
            return null;
        });
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/requests/AddTagRequest.kt
package y2k.joyreactor.services.requests

import y2k.joyreactor.common.async.CompletableContinuation
import y2k.joyreactor.common.async.runAsync
import y2k.joyreactor.common.http.HttpClient
import java.net.URLEncoder

/**
 * Created by y2k on 19/10/15.
 */
class AddTagRequest(private val httpClient: HttpClient) {

    fun request(tagName: String): CompletableContinuation<*> {
        return runAsync {
            val tagUrl = "http://joyreactor.cc/tag/" + URLEncoder.encode(tagName)
            val tagPage = httpClient.getDocument(tagUrl)
            val addTagLink = tagPage.select("a.change_favorite_link").first().absUrl("href")
            httpClient.getText(addTagLink)
        }
    }
}